internal/pkg/policy/policy_output.go

```diff
@@ -229,6 +233,35 @@ func (p *Output) prepareElasticsearch(
 		output.PermissionsHash = p.Role.Sha2 // for the sake of consistency
 	}
 
+	if outputMap.GetMap(p.Name).GetString("type") == OutputTypeRemoteElasticsearch {
```

Comment:
I might be missing something, but wouldn't [this switch block](https://github.com/elastic/fleet-server/pull/3051/files#diff-85caf77b270865c04855a90fc60468c90130cd83d2218aa66f362b32c178f4e0R57-R73) prevent `prepareElasticsearch` ever be called for `output.type == OutputTypeRemoteElasticsearch`?

Reply:
it is a bit hacky now as I've set `p.Type` to `elasticsearch` so the switch block goes to `prepareElasticsearch`. However the type in `outputMap` is still remote-elasticsearch.

Reply:
I see, I think the best is to get it all consistent, without this hack. Right now the way you're dealing with the remote-ES is different. So either it's be better to create another function to do all the handling of the remote-ES OR, my preferred solution, if they indeed share code, make the part that differs generic or configurable. For example the `generateOutputAPIKey` could be changed to take in a interface or ES client, so it can create the API key either on the fleet-ES or the remote-ES.

I see it's not the final version and you're still experimenting, but right now a lot of operation that do not serve the remote ES are performed before the remote-ES specific part which will indeed _prepare~Remote~Elasticsearch_.

Reply:
> For example the generateOutputAPIKey could be changed to take in a interface or ES client, so it can create the API key either on the fleet-ES or the remote-ES.

This is a good idea, hopefully then we can use the same `prepareElasticsearch` with a different es client.

I was looking at creating another bulker for remote ES but I think it's an overkill, as bulker does a lot of other things, and has to be started as a service.
I think it would be simpler for the existing bulker to store a list of remote ES clients (one per output) and use the ES client per output to generate api key.

Reply:
Made some changes in the latest commit, storing remote es clients in bulker to reuse between Prepare calls. It should probably be watching for hosts or service_token changes, to update the client.

Using remote-elasticsearch type, only changing it back to elasticsearch before sending to agent.

This way we can use the same prepareElasticsearch function, and only differ when an api key is created (and updated).
Tested that the functionality still works, and now the correct remote `api_key_id` is stored in the agent doc.

What's missing is using the remote es client when update api key is needed (it seems to be not that straighforward in bulker as create), and when api key is validated or invalidated.

Reply:
I'm looking at the code on how to create and run a new bulker, though it seems like a core logic to fleet server. 
Do I have to copy all of this code to create a new bulker? Probably `bulker.Run` and `runSubsystems` is needed too.
Maybe this code should be moved out to a separate function to decouple from the rest of the fleet server run.

https://github.com/elastic/fleet-server/blob/475e3ee0acdc0b9e33eee60cea1e5cabe91ced64/internal/pkg/server/fleet.go#L363-L417